### PR TITLE
ignore pruned dependencies and upgrade to github.com/distribution/distribution/v3

### LINKED ISSUE #103849: Migrate from github.com/docker/distribution to github.com/distribution/distribution/v3
#### What happened:
Kubernetes libraries have a dependency on `github.com/docker/distribution`
#### What you expected to happen:
Kubernetes libraries have a dependency on `github.com/distribution/distribution/v3`. Note the github org just changed, its the same project. I believe the v2 -> v3 was mostly to get around go.mod weirdness with versions, rather than large breaking changes, but am not sure.
#### How to reproduce it (as minimally and precisely as possible):
Look at go.mod, I suppose

#### Anything else we need to know?:
`github.com/docker/distribution@v2.7.1+incompatible`, the library in use, is 2.5 years old. It does not have a go.mod file, and in turn depends on a bunch of other libraries without go.mod files. This generally makes importing these libraries a pain - you can see 350 projects needing to add `replace` hacks to go.mod to work around these issues: https://github.com/search?q=%22replace+github.com%2Fdocker%2Fdistribution%22&type=code

It would be great to upgrade to the latest version.

### PR DESCRIPTION
**Pull request by Jamstah**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Following cleanup in distribution, this no longer adds extra transitive dependencies, so reopening based on https://github.com/kubernetes/kubernetes/pull/103850

Replace docker/distribution with distribution/distribution/v3. See https://github.com/kubernetes/kubernetes/issues/103849 for motivation

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/103849

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by Jamstah** [comment]
I'm not sure what the e2e-capz-windows-containerd failures are, perhaps flakes?

The dependencies:
``` 2023/03/21 20:17:50 Unwanted module "github.com/gorilla/handlers" marked in /home/prow/go/src/k8s.io/kubernetes/hack/unwanted-dependencies.json is referenced
2023/03/21 20:17:50    github.com/distribution/distribution/v3
2023/03/21 20:17:50 !!! Avoid updating referencing modules to versions that reintroduce use of unwanted dependencies
2023/03/21 20:17:50 Unwanted module "github.com/gorilla/mux" marked in /home/prow/go/src/k8s.io/kubernetes/hack/unwanted-dependencies.json is referenced
2023/03/21 20:17:50    github.com/distribution/distribution/v3
2023/03/21 20:17:50 !!! Avoid updating referencing modules to versions that reintroduce use of unwanted dependencies
2023/03/21 20:17:50 Unwanted module "github.com/hashicorp/golang-lru" marked in /home/prow/go/src/k8s.io/kubernetes/hack/unwanted-dependencies.json is referenced
2023/03/21 20:17:50    github.com/distribution/distribution/v3
2023/03/21 20:17:50 !!! Avoid updating referencing modules to versions that reintroduce use of unwanted dependencies 
```

While these are all dependencies of github.com/distribution/distribution/v3, they aren't used by the included package, so they don't get added to the go dependencies. Is the check working correctly? Have I missed something?

**Comment by Jamstah** [comment]
I've added a second commit to this that uses `go mod why` to check unwanted dependencies haven't been pruned, which has the side effect of removing a couple more unwanted dependencies from the exception list.

**Comment by Jamstah** [comment]
I've added a third commit to ignore dependencies that are used by tests in vendored packages, which gets rid of github.com/getsentry/raven-go from the unwanted dependency exceptions as well.

This PR now has two purposes - upgrading a package and changing how unwanted dependencies are detected. Let me know if I should split this out into two prs. A dependency process PR would probably drag in less reviewers :)

**Comment by Jamstah** [comment]
@liggitt @pacoxu FYI here because you worked on the dependency verifier previously.

File: cmd/dependencyverifier/dependencyverifier.go [hunk 1]
```diff
@@ -104,6 +104,27 @@ func convertToMap(modStr string) ([]module, map[module][]module) {
 	return mainModulesList, modMap
 }
 
+func getShortestPath(mod module) (path []string) {
+	modWhyStr, err := runCommand("go", "mod", "why", "-vendor", "-m", mod.name)
```
Comment (liggitt) [inline cmd/dependencyverifier/dependencyverifier.go:108]: I have found `go mod why` does not give accurate answers... I don't think we can use this for determining whether we have a path to a package/module

**Comment by liggitt** [comment]
> changing how unwanted dependencies are detected

I don't think I agree with this change... `github.com/distribution/distribution/v3` does actually add more paths to unwanted dependencies:

```
go mod graph | grep github.com/gorilla/mux
github.com/distribution/distribution/v3@v3.0.0-20230223072852-e5d5810851d1 github.com/gorilla/mux@v1.8.0
```

Dependency pruning is nice to make less stuff get fetched to build a given snapshot, but we _really_ don't want these unwanted dependencies in our graph _at all_.

The fact that we don't *happen* to currently depend on a package that pulls them into our build isn't enough. We're only one package import away (either in this repo or in the dependency's repo where we have less visibility/control) from pulling in the unwanted source, and the transitive dependencies of the pruned-but-unwanted dependency still inform the module graph and minimum version selection.

**Comment by Jamstah** [comment]
Understood. I'll not pursue that path then, which means we're blocked on closing #103849 until there is a reference module, not just a package. Or by adding more unwanted dependencies to the exceptions, which is unwarranted here really.

I'll close this, and look at splitting reference out of distribution instead.

**Comment by liggitt** [comment]
if we're only depending on the image spec parsing, I'd also consider a careful extraction of that code to a kubernetes-specific package. I honestly would prefer to have that code in a package we know won't evolve / change in the future in ways that would expand the surface area of what we accept for image spec values in the Kubernetes API

**Pull request closed without merging**